src/ir_pattern.py

```diff
@@ -0,0 +1,443 @@
+            return False
+        if not all(self.match_access(*accs) for accs in zip(pat.inputs, cand.inputs)):
+            return False
+        if not all(self.match_access(*accs) for accs in zip(pat.outputs, cand.outputs)):
+            return False
+        # TODO: Do clobbers also need to be matched?
+        return True
+
+    def match_ref(self, pat: Reference, cand: Reference) -> bool:
+        if isinstance(pat, str) and isinstance(cand, str):
+            return pat == cand
+        return self._match_var(self.ref_map, pat, cand)
+
+    def match_refset(self, pat: RefSet, cand: RefSet) -> bool:
+        if len(pat) > len(cand):
+            return False
+        # TODO: This may need backtracking?
+        refs = cand.copy()
+        for e in pat:
+            assert len(refs) >= 1
+            for a in refs:
+                if self.match_ref(e, a):
+                    refs.remove(a)
+                    break
+            else:
+                return False
+        return True
+
+    def match_accessrefs(self, pat: AccessRefs, cand: AccessRefs) -> bool:
+        for pat_reg, pat_refs in pat.items():
+            # For now, skip mapping any memory accesses outside of the stack.
+            # Usually, these accesses are not intended to be part of the matched pattern,
+            # but in the future the pattern syntax could be extended to explicitly mark
+            # which accesses must be matched.
+            if (
+                isinstance(pat_reg, MemoryAccess)
+                and pat_reg.base_reg != self.arch.stack_pointer_reg
+            ):
+                continue
+            cand_reg = self.map_access(pat_reg)
+            cand_refs = cand.get(cand_reg)
+            if not self.match_refset(pat_refs, cand_refs):
+                return False
+        return True
+
+
+def simplify_ir_patterns(
+    arch: ArchFlowGraph, flow_graph: FlowGraph, pattern_classes: List[Type[IrPattern]]
+) -> None:
+    # Precompute a RefSet for each mnemonic
+    refs_by_mnemonic = defaultdict(list)
+    for node in flow_graph.nodes:
+        for i, instr in enumerate(node.block.instructions):
+            ref = InstrRef(node, i)
+            refs_by_mnemonic[instr.mnemonic].append(ref)
+
+    def replace_instr(ref: InstrRef, new_asm: AsmInstruction) -> None:
+        # Remove ref from all instr_references
+        instr = ref.instruction()
+        for rs in flow_graph.instr_inputs[ref].values():
+            for r in rs:
+                if isinstance(r, InstrRef):
+                    flow_graph.instr_references[r].remove_ref(ref)
+
+        # Parse the asm & set the clobbers
+        new_instr = arch.parse(new_asm.mnemonic, new_asm.args, instr.meta.derived())
+        new_instr.clobbers.extend(
+            acc for acc in instr.outputs if acc not in new_instr.outputs
+        )
+        new_instr.clobbers.extend(
+            acc for acc in instr.clobbers if acc not in new_instr.clobbers
+        )
```

Comment:
we're clobbering a lot of registers here. What if they are read from? That should make us not do the replacement, right?

Reply:
There is a check near line 426 for when the instruction is part of the pattern but will not be replaced (because its outputs are referenced by other non-pattern instructions): if the instruction would clobber an input, then the entire replacement is cancelled.
```python
                elif any(r in pattern_inputs for r in instr.clobbers + instr.outputs):
                    # If this instruction can't be replaced, but it clobbers a needed
                    # input register, then we can't perform the rewrite
```

However, in the more common case, I think there is an issue here: if the instruction clobbers an input register *and* is being replaced, the replaced `nop` should not be marked as clobbering the register. I think the current code is incorrect, but still behaves correctly because the register flow graph is not recomputed from scratch after the rewrites.